Sources/Task/TaskProgress.swift

```diff
@@ -29,87 +29,145 @@ private struct KVO {
+            observee.removeObserver(self, forKeyPath: KeyPath.paused.rawValue, context: &Observation.pausedContext)
         }
 
-        if #available(OSX 10.11, iOS 9.0, *) {
-            resumingHandler = original.resume
+        private override func observeValueForKeyPath(keyPath: String?, ofObject object: AnyObject?, change: [String : AnyObject]?, context: UnsafeMutablePointer<Void>) {
+            guard let keyPath = keyPath where object != nil && (state & State.ready.rawValue) != 0, let observer = observer, newValue = change?[NSKeyValueChangeNewKey] else { return }
+            switch context {
+            case &Observation.cancelledContext:
+                observer.inheritCancelled(newValue as! Bool)
+            case &Observation.pausedContext:
+                observer.inheritPaused(newValue as! Bool)
+            case &Observation.attributesContext:
+                observer.inheritValue(newValue, forKeyPath: keyPath)
+            default:
+                preconditionFailure("Unexpected KVO context for private object")
+            }
         }
     }
 
-    func detach() {
-        for keyPath in KVO.KeyPath.all {
-            original.removeObserver(self, forKeyPath: keyPath.rawValue, context: &KVO.context)
+    let observee: NSProgress
+    var token: Observation?
+
+    init(attachingTo observee: NSProgress) {
+        let current = NSProgress.currentProgress()
+        self.observee = observee
+        super.init(parent: current, userInfo: nil)
+
+        if current?.cancelled == true {
+            observee.cancel()
+        }
+
+        if current?.paused == true {
+            observee.pause()
         }
+
+        token = Observation(observing: observee, observer: self)
     }
 
-    override func observeValueForKeyPath(keyPath: String?, ofObject object: AnyObject?, change: [String : AnyObject]?, context: UnsafeMutablePointer<Void>) {
-        switch (keyPath, context) {
-        case (KVO.KeyPath.cancelled.rawValue?, &KVO.context):
-            guard change?[NSKeyValueChangeNewKey] as? Bool == true else { return }
+    deinit {
+        token?.cancelObserving(observee)
+        token = nil
+    }
+
+    func inheritCancelled(value: Bool) {
+        if value {
             cancellationHandler = nil
             cancel()
-        case (KVO.KeyPath.paused.rawValue?, &KVO.context):
-            if change?[NSKeyValueChangeNewKey] as? Bool == true {
-                pausingHandler = nil
-                pause()
-                if #available(OSX 10.11, iOS 9.0, *) {
-                    resumingHandler = original.resume
-                }
-            } else if #available(OSX 10.11, iOS 9.0, *) {
-                resumingHandler = nil
-                resume()
-                pausingHandler = original.pause
+        } else {
+            cancellationHandler = observee.cancel
+        }
+    }
+
+    func inheritPaused(value: Bool) {
+        if value {
+            pausingHandler = nil
```

Comment:
Yes, and a couple of other reasons, like cleaning up potential cycles from cancellation and to prevent recursively calling the handler.